test/reader.js

```diff
@@ -0,0 +1,328 @@
+        reader('?!?', ipfs._ipldResolver),
+        collect((error) => {
+          expect(error.message).to.contain('Non-base58 character')
+
+          done()
+        })
+      )
+    })
+
+    it('fails on non-file', (done) => {
+      ipfs.files.add({
+        path: '/my-directory'
+      }, (error, result) => {
+        if (error) {
+          return done(error)
+        }
+
+        pull(
+          reader(result[0].hash, ipfs._ipldResolver),
+          collect((error) => {
+            expect(error.message).to.contain('was not a file')
+
+            done()
+          })
+        )
+      })
+    })
+
+    it('fails on file with no links', (done) => {
+      ipfs.files.add({
+        path: '/my-file',
+        content: Buffer.from([])
+      }, (error, result) => {
+        if (error) {
+          return done(error)
+        }
+
+        pull(
+          reader(result[0].hash, ipfs._ipldResolver),
+          collect((error) => {
+            expect(error.message).to.contain('had no links')
+
+            done()
+          })
+        )
+      })
+    })
+
+    it('reads bytes with a begin', (done) => {
+      ipfs.files.add({
+        path: '/my-file',
+        content: Buffer.from([0, 1, 2, 3])
+      }, (error, result) => {
+        if (error) {
+          return done(error)
+        }
+
+        pull(
+          reader(result[0].hash, ipfs._ipldResolver, 1),
+          collect((error, results) => {
+            if (error) {
+              return done(error)
+            }
+
+            const data = Buffer.concat(results)
+
+            expect(data).to.deep.equal(Buffer.from([1, 2, 3]))
+
+            done()
+          })
+        )
+      })
+    })
+
+    it('reads bytes with a negative begin', (done) => {
+      ipfs.files.add({
+        path: '/my-file',
+        content: Buffer.from([0, 1, 2, 3])
+      }, (error, result) => {
+        if (error) {
+          return done(error)
+        }
+
+        pull(
+          reader(result[0].hash, ipfs._ipldResolver, -1),
+          collect((error, results) => {
+            if (error) {
+              return done(error)
+            }
+
+            const data = Buffer.concat(results)
+
+            expect(data).to.deep.equal(Buffer.from([3]))
+
+            done()
+          })
+        )
+      })
+    })
+
+    it('reads bytes with an end', (done) => {
+      ipfs.files.add({
+        path: '/my-file',
+        content: Buffer.from([0, 1, 2, 3])
+      }, (error, result) => {
+        if (error) {
+          return done(error)
+        }
+
+        pull(
+          reader(result[0].hash, ipfs._ipldResolver, 0, 1),
+          collect((error, results) => {
+            if (error) {
+              return done(error)
+            }
+
+            const data = Buffer.concat(results)
+
+            expect(data).to.deep.equal(Buffer.from([0]))
+
+            done()
+          })
+        )
+      })
+    })
+
+    it('reads bytes with a negative end', (done) => {
+      ipfs.files.add({
```

Comment:
Everything except the assertion in the end seems duplicated across the test cases. Could you extract them into a function and call that instead please?

Reply:
Done!